File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    private boolean dfs(
        Map<Integer, List<Integer>> finishToTakeMap,
        Integer finish,
        List<Integer> takes,
        List<Integer> taken
    ) {
        if (takes.contains(finish)) return false;

        if (taken.contains(finish)) return true;

        if (finishToTakeMap.containsKey(finish)) {
            takes.add(finish);

            for (Integer take : finishToTakeMap.get(finish)) {
                if (!dfs(finishToTakeMap, take, takes, taken)) return false;
            }

            takes.remove(finish);
            taken.add(finish);
        }

        return true;
    }

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> finishToTakeMap = new HashMap<>();
        for (int[] prerequisite : prerequisites) {
            finishToTakeMap.putIfAbsent(prerequisite[0], new ArrayList<>());
            finishToTakeMap.get(prerequisite[0]).add(prerequisite[1]);
        }

        List<Integer> takes = new ArrayList<>();
        List<Integer> taken = new ArrayList<>();
        for (Integer finish : finishToTakeMap.keySet()) {
            if (!dfs(finishToTakeMap, finish, takes, taken)) return false;
        }

        return true;
    }
}

File: 0207-course-schedule/0207-course-schedule.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val finishToTakeMap: MutableMap<Int, MutableList<Int>> = mutableMapOf()
        for (prerequisite in prerequisites) {
            finishToTakeMap.putIfAbsent(prerequisite[0], mutableListOf())
            finishToTakeMap[prerequisite[0]]!!.add(prerequisite[1])
        }

        val takes: MutableList<Int> = mutableListOf()
        val taken: MutableList<Int> = mutableListOf()

        fun dfs(finish: Int, takes: MutableList<Int>, taken: MutableList<Int>): Boolean {
            if (takes.contains(finish)) return false

            if (taken.contains(finish)) return true

            if (finishToTakeMap.containsKey(finish)) {
                takes.add(finish)

                for (take in finishToTakeMap[finish]!!) {
                    if (!dfs(take, takes, taken)) return false
                }

                takes.remove(finish)
                taken.add(finish)
            }

            return true
        }

        for (finish in finishToTakeMap.keys) {
            if (!dfs(finish, takes, taken)) return false
        }

        return true
    }
}